src/app/wallets/notify-on-chain-receipt-unconfirmed.ts

```diff
@@ -0,0 +1,105 @@
+import { BTC_NETWORK, ONCHAIN_SCAN_DEPTH } from "@config"
+
+import { getCurrentPrice } from "@app/prices"
+
+import { CacheKeys } from "@domain/cache"
+import { TxDecoder } from "@domain/bitcoin/onchain"
+import { DepositFeeCalculator } from "@domain/wallets"
+import { DisplayCurrency, DisplayCurrencyConverter } from "@domain/fiat"
+import { CouldNotFindWalletFromOnChainAddressesError } from "@domain/errors"
+
+import {
+  AccountsRepository,
+  UsersRepository,
+  WalletsRepository,
+} from "@services/mongoose"
+import { RedisCacheService } from "@services/cache"
+import { OnChainService } from "@services/lnd/onchain-service"
+import { NotificationsService } from "@services/notifications"
+
+export const notifyOnChainReceiptUnconfirmedByTxHash = async ({
+  txHash,
+  scanDepth = ONCHAIN_SCAN_DEPTH,
+  logger,
+}: {
+  txHash: OnChainTxHash
+  scanDepth?: ScanDepth
+  logger: Logger
+}): Promise<true | ApplicationError> => {
+  RedisCacheService().clear({ key: CacheKeys.LastOnChainTransactions })
+
+  const onChainService = OnChainService(TxDecoder(BTC_NETWORK))
+  if (onChainService instanceof Error) return onChainService
+
+  const tx = await onChainService.lookupOnChainReceipt({ txHash, scanDepth })
+  if (tx instanceof Error) return tx
+  // we only handle pending notification here because we wait more than 1 block
+  if (tx.confirmations > 0) return true
+
+  logger.info({ transactionType: "receipt", pending: true }, "mempool appearance")
+
+  const addresses = tx.uniqueAddresses()
+  const wallets = await WalletsRepository().listByAddresses(addresses)
+  // we should not process hot wallet rebalances
+  if (wallets instanceof CouldNotFindWalletFromOnChainAddressesError) return true
+  if (wallets instanceof Error) return wallets
+
+  for (const wallet of wallets) {
+    const walletId = wallet.id
+    logger.trace({ walletId, txHash }, "notifying onchain receipt")
+
+    const result = await processTxForWallet({ wallet, tx })
+    if (result instanceof Error) {
+      logger.error({ walletId, txHash, error: result }, "error notifying onchain receipt")
+    }
+  }
+
+  return true
+}
+
+const processTxForWallet = async ({
+  wallet,
+  tx,
+}: {
+  wallet: Wallet
+  tx: IncomingOnChainTransaction
+}): Promise<void | ApplicationError> => {
+  const displayCurrencyPerSat = await getCurrentPrice()
+  if (displayCurrencyPerSat instanceof Error) return displayCurrencyPerSat
+
+  const converter = DisplayCurrencyConverter(displayCurrencyPerSat)
+  const account = await AccountsRepository().findById(wallet.accountId)
+  if (account instanceof Error) return account
+
+  const walletAddresses = wallet.onChainAddresses()
+  for (const { sats, address } of tx.rawTx.outs) {
```

Comment:
Just noting that a hypothetical spam side-effect here is recipient could have something happen like receiving multiple notifications for a single transaction if someone for e.g. made a txn with 100 utxos all with some small sat amount paid to the same address.

It's not too likely because there is a cost attached, but it's still good to be aware of.

Reply:
uhmm yep but still we need to do it in this way for final settlement but... you are right, there is a cost to do that so I think is not a problem